crates/harp/src/envvar.rs

```diff
@@ -0,0 +1,109 @@
+//
+// envvar.rs
+//
+// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
+//
+//
+
+use crate::protect::RProtect;
+use crate::r_lang;
+use crate::r_string;
+use crate::r_symbol;
+use crate::RObject;
+
+/// Set an environment variable through `Sys.setenv()`
+///
+/// This is a safe alternative to [std::env::set_var()]. On Windows in particular,
+/// using [std::env::set_var()] after R has started up (i.e. after Ark opens the
+/// R DLL and calls `setup_Rmainloop()`) has no effect. We believe this is because
+/// [std::env::set_var()] ends up calling `SetEnvironmentVariableW()` to set the
+/// environment variable, which only affects the Windows API environment space and
+/// not the C environment space that R seems to have access to after it has started
+/// up. This matters because R's `Sys.getenv()` uses the C level `getenv()` to
+/// access environment variables, which only looks at C environment space.
+///
+/// To affect the C environment space, [std::env::set_var()] would have had to call
+/// [libc::putenv()], but that has issues with thread safety, and we have seen this
+/// crash R before, so we'd like to stay away from using that ourselves.
+///
+/// The easiest solution to this problem is to just go through R's `Sys.setenv()`
+/// to ensure that R picks up the environment variable update.
+///
+/// If R has not started up yet, you should be safe to call [std::env::set_var()].
+/// For example, we do this for `R_HOME` during the startup process and for
+/// `R_PROFILE_USER` in some tests.
```

Comment:
This means the C runtime environment variables are synchronised with the Windows API ones during the R startup (though we don't know exactly how).